ValveResourceFormat/IO/MaterialExtract.cs

```diff
@@ -0,0 +1,482 @@
+            if (texture is null)
+            {
+                continue;
+            }
+
+            var images = GetInputImagesForTexture(type, filePath, true, material.IntParams, material.VectorParams);
+            var vtex = new TextureExtract(texture).ToMaterialMaps(images);
+
+            vmat.ExternalRefsHandled.Add(filePath + "_c", vtex);
+            vmat.SubFilesAreExternal = true;
+            vmat.SubFiles.AddRange(vtex.SubFiles);
+        }
+
+        return vmat;
+    }
+
+    public static string OutTextureName(string texturePath, bool keepOriginalExtension, string desiredSuffix = null)
+    {
+        if (!texturePath.EndsWith(".vtex", StringComparison.OrdinalIgnoreCase))
+        {
+            throw new ArgumentException($"{texturePath} must have .vtex", nameof(texturePath));
+        }
+        texturePath = Path.ChangeExtension(texturePath, null); // strip .vtex
+        texturePath = Path.ChangeExtension(texturePath, null); // strip .generated (if present)
+
+        var textureParts = Path.GetFileName(texturePath).Split('_');
+        if (textureParts.Length > 2)
+        {
+            if (textureParts[^1].All("0123456789abcdef".Contains)             // This is a hash
+            && (textureParts[^2].Length >= 3 && textureParts[^2].Length <= 4) // This is the original extension
+            && !string.IsNullOrEmpty(string.Join("_", textureParts[..^2])))   // Name of the Input[0] image
+            {
+                texturePath = Path.Combine(Path.GetDirectoryName(texturePath), string.Join("_", textureParts[..^2]));
+                texturePath = texturePath.Replace(Path.DirectorySeparatorChar, '/');
+            }
+            // Also seen "{MAT}_vmat_{G_TPARAM}_{HASH}.vtex"
+        }
+
+        if (desiredSuffix is not null)
+        {
+            // Strip e.g. _color if we want _metal
+            if (!texturePath.EndsWith(desiredSuffix, StringComparison.OrdinalIgnoreCase))
+            {
+                foreach (var suffix in CommonTextureSuffixes.Values)
+                {
+                    if (texturePath.EndsWith(suffix, StringComparison.OrdinalIgnoreCase))
+                    {
+                        texturePath = texturePath[..^suffix.Length];
+                        break;
+                    }
+                }
+            }
+
+            // Keep the hash to avoid conflicts
+            texturePath += "_" + textureParts[^1] + desiredSuffix;
+        }
+
+        return Path.ChangeExtension(texturePath, keepOriginalExtension ? textureParts[^2] : "png");
+    }
+
+    public IEnumerable<UnpackInfo> GetInputImagesForTexture(string textureType, string texturePath, bool ommitDefault,
+            Dictionary<string, long> intParams, Dictionary<string, Vector4> vectorParams)
+    {
+        var shader = material.ShaderName[..^4];
+
+        if (TextureMappings.ContainsKey(shader) && TextureMappings[shader] is IDictionary<string, ChannelMapping[]> mappings && mappings.ContainsKey(textureType))
```

Comment:
Use TryGetValue, and why the `is` check? When can it possibly be not that.

Reply:
Fixed!